bbpower/compsep.py

```diff
@@ -1,239 +1,364 @@
+        _,_,_,self.ell_b,_ = self.order[0]
+        self.n_bpws = len(self.ell_b)
+        self.windows = np.zeros([self.ncross, self.n_bpws, len(self.bpw_l)])
+
+        #Get power spectra and covariances
+        v = self.s.mean.vector
+        if len(v) != self.n_bpws * self.ncross:
+            raise ValueError("C_ell vector's size is wrong")
+        cv = self.s.precision.getCovarianceMatrix()
 
+        #Parse into the right ordering
+        v2d = np.zeros([self.n_bpws, self.ncross])
+        if self.use_handl:
+            v2d_noi = np.zeros([self.n_bpws, self.ncross])
+            v2d_fid = np.zeros([self.n_bpws, self.ncross])
+        cv2d = np.zeros([self.n_bpws, self.ncross, self.n_bpws, self.ncross])
+        self.vector_indices = self.vector_to_matrix(np.arange(self.ncross, dtype=int)).astype(int)
         self.indx = []
         for t1,t2,typ,ells,ndx in self.order:
-            if typ == b'BB':
-                self.indx+=list(ndx)
+            for b,i in enumerate(ndx):
+                self.windows[self.vector_indices[t1, t2], b, :] = self.s.binning.windows[i].w
+            v2d[:,self.vector_indices[t1, t2]] = v[ndx]
+            if self.use_handl:
+                v2d_noi[:, self.vector_indices[t1, t2]] = s_noi.mean.vector[ndx]
+                v2d_fid[:, self.vector_indices[t1, t2]] = s_fid.mean.vector[ndx]
+            if len(ells) != self.n_bpws:
+                raise ValueError("All power spectra need to be sampled at the same ells")
+            for t1b, t2b, typb, ellsb, ndxb in self.order:
+                cv2d[:, self.vector_indices[t1, t2], :, self.vector_indices[t1b, t2b]] = cv[ndx, :][:, ndxb]
 
-        self.bbdata = self.data[self.indx]
-        self.bbcovar = self.covar[self.indx][:, self.indx]
+        #Store data
+        self.bbdata = self.vector_to_matrix(v2d)
+        if self.use_handl:
+            self.bbnoise = self.vector_to_matrix(v2d_noi)
+            self.bbfiducial = self.vector_to_matrix(v2d_fid)
+        self.bbcovar = cv2d.reshape([self.n_bpws * self.ncross, self.n_bpws * self.ncross])
         self.invcov = np.linalg.solve(self.bbcovar, np.identity(len(self.bbcovar)))
-        # check this inverse is good? 
-
-        # Load CMB B-mode data
-        # TODO: Incorporate loading into the pipeline?
-        # check units 
-        # strong assumption that this file is sampled at the same ells as the bandpass_l (which is true for now)
-        # otherwise we will have to interpolate. That's yucky though so better to enforce this sampling. 
-        cmb_bbfile = np.loadtxt('/Users/abitbol/code/self_calibration_fg/data/camb_lens_r1.dat')
-        cmb_lensingfile = np.loadtxt('/Users/abitbol/code/self_calibration_fg/data/camb_lens_nobb.dat')
+        return
+
+    def load_cmb(self):
+        """
+        Loads the CMB BB spectrum as defined in the config file. 
+        """
+        cmb_lensingfile = np.loadtxt(self.config['cmb_files'][0])
+        cmb_bbfile = np.loadtxt(self.config['cmb_files'][1])
```

Comment:
CAMB files usually start at ell=2. Are we taking care of this?

Reply:
So I manually added a line of zeros in these files to match with the BICEP files (I think just the window files) which start at ell=1. We should probably just let them start at ell=2 and have a line of code that either strips off the ell=1 data from the BICEP stuff or append zeros to the top of the CMB file. What do you think?

Reply:
Sounds good, let's go with that